### PR TITLE
[JAX] Debugging inspect utility

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Given `jax.debug.print/callback` is currently broken (issue), this PR introduces an experimental alternative for use for our own internal debugging. This new debugging API allows us to inspect tensors but will be experimental and may have breaking changes without a deprecation process.

Usage:
```
     x = <some logic to compute x>

      from transformer_engine.jax.debug.experimental import inspect_array as te_inspect_array
      x= te_inspect_array(x, "some_name")

     <something that consumes  x>
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Introduces a new debugging tool for dumping binary blobs of tensor values for multi-GPU inspection.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/csrc/extensions/inspect.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  // Log the tensor metadata to the console
+  printf("Tensor data written to %s (shape: [", filename.c_str());
+  for (size_t i = 0; i < input_buf.dimensions().size(); ++i) {
+    printf("%zu", static_cast<size_t>(input_buf.dimensions()[i]));
+    if (i < input_buf.dimensions().size() - 1) {
+      printf(", ");
+    }
+  }
+  printf("], dtype: %d", static_cast<int>(input_buf.element_type()));
+  printf(", min: %f, max: %f, mean: %f, std: %f)\n", min_val, max_val, mean_val, std_val);
```
Comment (tdophung) [inline transformer_engine/jax/csrc/extensions/inspect.cpp:82]: If you want to keep this print, it would also be nice to print out where the stat is coming from (which GPU/shard)

File: transformer_engine/jax/debug/experimental/inspect.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        filename (str): The path to the binary file containing the array data.
+        shape (tuple): The shape of the array to be loaded.
+        dtype (jnp.dtype): The data type of the array to be loaded.
+
+    Returns:
+        jnp.ndarray: The loaded JAX array.
+    """
+    with open(filename, "rb") as f:
+        data = f.read()
+    array = jnp.frombuffer(data, dtype=dtype).reshape(shape)
```
Comment (tdophung) [inline transformer_engine/jax/debug/experimental/inspect.py:173]: Does this handle 4-bit data as well? Will the data printed be legible for e2m1 or e0m3?
This might be useful for debugging quantization outputs

File: transformer_engine/jax/csrc/extensions/inspect.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  std::vector<uint8_t> input_data(input_buf.size_bytes());
+  cudaMemcpyAsync(input_data.data(), input_buf.untyped_data(), input_buf.size_bytes(),
+                  cudaMemcpyDeviceToHost, stream);
+
+  float min_val{}, max_val{}, mean_val{}, std_val{};
+  cudaMemcpyAsync(&min_val, min_buf.untyped_data(), sizeof(float), cudaMemcpyDeviceToHost, stream);
+  cudaMemcpyAsync(&max_val, max_buf.untyped_data(), sizeof(float), cudaMemcpyDeviceToHost, stream);
+  cudaMemcpyAsync(&mean_val, mean_buf.untyped_data(), sizeof(float), cudaMemcpyDeviceToHost,
+                  stream);
+  cudaMemcpyAsync(&std_val, std_buf.untyped_data(), sizeof(float), cudaMemcpyDeviceToHost, stream);
```
Comment (tdophung) [inline transformer_engine/jax/csrc/extensions/inspect.cpp]: Should do error checking here. I think it is NVTE_CHECK_CUDA?

File: transformer_engine/jax/debug/experimental/inspect.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    )
+
+
+@partial(jax.custom_vjp, nondiff_argnums=())
+def _inspect(
+    x,
+):
+    """ """
+    output, _ = _inspect_fwd_rule(
+        x,
```
Comment (tdophung) [inline transformer_engine/jax/debug/experimental/inspect.py:123]: is this for eventually inspecting grads? Maybe we can put in a TODO here for inspecting grads?

File: transformer_engine/jax/debug/experimental/inspect.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+
+def inspect_array(x: jnp.ndarray, name: str) -> jnp.ndarray:
+    """Utility function to inspect JAX arrays by printing their name, shape, dtype, and statistics.
+
+    Args:
+        x (jnp.ndarray): The JAX array to inspect.
+        name (str): The name of the array for identification in the output.
+    """
+    # TODO: Handle the name of the tensor in the primitive and output files
```
Comment (tdophung) [inline transformer_engine/jax/debug/experimental/inspect.py]: This does not need to be in this PR, but I did ask claude about this and it suggested 2 options:
1. We can hash the string name into a number, to pass to XLA, and maintain a hash map on the python side to attach to the filename. 
2. We do not pass the name to _inspect (like what you are doing here), but we "mv" the file in the filesystem to a different name depending on the argument `name`
Wdyt?

Comment (jberchtold-nvidia) [reply transformer_engine/jax/debug/experimental/inspect.py]: I'm not sure if option 2 will work, since we are running the Python "inspect" function during trace time but the file doesn't get written until the JAX program is jit'd and actually running. So I'm not sure how we'd move the file unless we did something like in option 1 with a dictionary.

I do like option 1, that is a good idea and may be our best option. Another alternative I was considering but haven't looked into yet is a static buffer with a fixed maximum string capacity. On the attention side we pass static args thru a buffer that is a struct that can support decoding fields on the C++ side. There may be a way to pack the  string on the Python side and read on the  C++ side assuming the string doesn't exceed some fixed char array maxsize. But that idea still needs some investigation, so your option 1 sounds like the most promising.

File: transformer_engine/jax/debug/experimental/inspect.py (hunk 4 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/debug/experimental/inspect.py:123]: This isn't directly for inspecting gradients. If we insert a custom primitive into a JAX graph, JAX errors since it doesn't understand that our inspect FFI is really a no-op and therefore it doesn't know how to compute the gradient of it. So this VJP rule just helps JAX know how to handle this in the backward and pass gradients thru.

This current inspect API wraps a single tensor, which makes sense for the forward. We just wrap some intermediate result in `inspect` and we can look at the results. Usually this would be immediately after some custom operation we added, e.g. comparing MaxText vs. TE permute. However, since we're just wrapping the tensor, the gradient in this VJP will actually be the incoming gradient from the following operation, not the gradient of our custom op.

So I think we do want support for inspecting gradients, but instead of wrapping individual tensors  on the forward, it may make more sense to wrap a given function, like  our permute, and let us inspect the forward outputs and gradient outputs from our op (not the incoming gradients)

I have a somewhat hard-coded version of this idea working on my branch that for a GEMM logs stats on fwd output and dgrad and wgrad. If we make it generic enough, something similar could be used for permute and other ops too

Wdyt?

File: transformer_engine/jax/csrc/extensions/inspect.cpp (hunk 3 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/csrc/extensions/inspect.cpp]: Good point, updated

File: transformer_engine/jax/csrc/extensions/inspect.cpp (hunk 1 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/csrc/extensions/inspect.cpp:82]: Good idea, added "[gpu%d]:" as a prefix here

File: transformer_engine/jax/debug/experimental/inspect.py (hunk 2 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/debug/experimental/inspect.py:173]: I tried and unfortunately it doesn't seem to work with fp4. I'm not sure if this is an issue with my FFI or if it's an issue with fp4 support in jnp.frombuffer. But I've added unit tests for the other dtypes with a note that fp4 doesn't work currently

Comment (tdophung) [reply transformer_engine/jax/debug/experimental/inspect.py:173]: Sounds good, we can figure out the expansion to 4-bit later

File: transformer_engine/jax/debug/experimental/inspect.py (hunk 5 above)
Comment (tdophung) [reply transformer_engine/jax/debug/experimental/inspect.py]: option 1 would likely be easier implement IMO since you do not have to try decoding the passed static buffer into multiple fields (I assume the struct might have multiple field?) . Anyways I will leave this here without resolving the thread to re-visit

File: transformer_engine/jax/debug/experimental/inspect.py (hunk 4 above)
Comment (tdophung) [reply transformer_engine/jax/debug/experimental/inspect.py:123]: that seems like a cool idea, but if there are multiple outputs, we need a way to differentiate them. and potentially save them to different bins.

File: transformer_engine/jax/csrc/extensions/inspect.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdophung) [reply transformer_engine/jax/csrc/extensions/inspect.cpp:82]: This is fine, we want to see this whenenver we use inspect util

**Comment by tdophung** [comment]
LGTM! Thanks for making this change
